File: leetcode/c++/0054-spiral-matrix.cc
// https://leetcode.com/problems/spiral-matrix/
// Medium

#include<vector>
using namespace std;

class Solution {
vector<vector<bool>> checked;
    vector<int> ret;
    int rows, cols;
    enum Direction {left, right, top, bottom};
    Direction direction = right;
    pair<int, int> d = {0, 1};

    void changeDirection() {
        if (direction == right) {
            direction = bottom;
            d = {1, 0};
        } else if (direction == bottom) {
            direction = left;
            d = {0, -1};
        } else if (direction == left) {
            direction = top;
            d = {-1, 0};
        } else {
            direction = right;
            d = {0, 1};
        }
    }

    bool spiralOrderRec(vector<vector<int>>& matrix, int i, int j) {
        if (i >= rows || j >= cols || i < 0 || j < 0 || checked[i][j]) return false;
        ret.push_back(matrix[i][j]);
        checked[i][j] = true;
        for (int k = 0; k < 4; ++k) {
            if (spiralOrderRec(matrix, i+d.first, j+d.second)) {
                return true;
            }
            changeDirection();
        }
        return true;
    }
    
    

public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        rows = matrix.size();
        cols = matrix[0].size();
        checked = vector<vector<bool>> (rows, vector<bool> (cols, false));
        spiralOrderRec(matrix, 0, 0);
        return ret;
    }
};

File: leetcode/python/0054-spiral-matrix.py
# https://leetcode.com/problems/spiral-matrix/
# Medium

# Have indexes that start with `top, left, bottom, right = 0, 0, len(matrix)-1, len(matrix[0])-1`
# Then increase top everytime you go right, decrease right when you go down and vice versa
# Your break condition is `top <= bottom and left <= right`
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        top, left, bottom, right = 0, 0, len(matrix)-1, len(matrix[0])-1
        move = 0
        ret = []
        
        while top <= bottom and left <= right:
            if move == 0:
                for i in range(left, right+1):
                    ret.append(matrix[top][i])
                top += 1
            elif move == 1:
                for i in range(top, bottom+1):
                    ret.append(matrix[i][right])
                right -= 1
            elif move == 2:
                for i in range(right, left-1, -1):
                    ret.append(matrix[bottom][i])
                bottom -= 1
            elif move == 3:
                for i in range(bottom, top-1, -1):
                    ret.append(matrix[i][left])
                left += 1
            move = (move + 1)%4
        return ret
            
        return ret